text/3366-diagnostic-attribute-namespace.md

```diff
@@ -0,0 +1,222 @@
+)]
+#[diagnostic::on_unimplemented(
+    if(any(A = i32, Self = i32)),
+    note = "That's emitted if A or Self is a i32",
+)]
+// this attribute will not have any affect as
+// the attribute above will always match before
+#[diagnostic::on_unimplemented(
+    if(all(A = i32, Self = i32)),
+    note = "That's emitted if A and Self is a i32"
+)]
+#[diagnostic::on_unimplemented(
+    if(not(A = String)),
+    // and implicitly `A` is not a `i32` as that case is
+    // matched earlier
+    note = "That's emitted if A is not a `String`"
+)]
+#[diagnostic::on_unimplemented(
+    message="message",
+    label="label",
+    note="That's emitted if neither of the condition above are meet",
+)]
+trait MyIterator<A> {
+    fn next(&mut self) -> A;
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+A possible drawback is that this feature adds additional complexity to the compiler implementation. The compiler needs to handle an additional attribute namespace with at least one additional attribute. 
+
+Another drawback is that crates might hint lower quality error messages than the compiler itself. Technically the compiler would be free to ignore such hints, practically I would assume that it is impossible to judge the quality of such error messages in an automated way.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This proposal tries to improve error messages generated by rustc. It would give crate authors a tool to influence what error message is emitted in a certain situation, as they might sometimes want to provide specific details on certain error conditions. Not implementing this proposal would result in the current status quo. Currently the compiler always shows a "general" error message, even if it would be helpful to show additional details.
+
+There are alternative designs for the proposed `on_unimplemented` attribute:
+
+* The `on()` based filtering might be replaceable by placing the attribute on negative trait impls. This would turn a filter like
+```rust
+#[diagnostic::on_unimplemented(
+    on(Self = `String`, message = "Strings do not implement `IntoIterator` directly")
+)]
+trait IntoIterator {}
+```
+
+into the following negative trait impl:
+```rust
+#[diagnostic::on_unimplemented(message = "Strings do not implement `IntoIterator` directly")]
+impl !IntoIterator for String {}
+```
+
+This would simplify the syntax of the proposed attribute, but in turn block the implementation of type based filtering on the stabilization of `negative_impls`. On the other hand it would likely simplify writing more complex filters, that match only a certain generic set of types and it would prevent "duplicating" the filter-logic as this reuses the exiting trait system. The large disadvantage of this approach is that it couples error messages to the crates public API. Removing a negative trait impl is a breaking change, removing a `#[on_unimplemented]` attribute is only a change in the emitted compiler error.
```

Comment:
You need to call out that the simplification of complex filtering logic also requires some level of specialization support. This would be fine because even if the current impl is unsound it doesn't affect this use case, but it might close doors and block future specialization behavior changes.

Reply:
That's a good point. Will add that :+1:

Reply:
Another thing is that doing the attribute approach now *doesn't* preclude us from implementing a more integrated system in the future.